# Enrichment culture dataset analysis

This directory contains the data and scripts used in the analysis for the enrichment culture dataset for the Outbreak project.

* `data` directory contains metadata and small output data files generated from the analysis
* `bin` directory contains scripts and workflows used in the analysis
* `testing_scripts` directory contains scripts in testing (not necessarily working)
* `docs` directory contains usage/information of scripts and data
* `archive` directory contains old and unused scripts and old data

### Scripts usage
-------------------------------------------------------
#### main scripts:
* [assembly_map_vcall.nf](docs/assembly_map_vcall.md) - for genome assembly using both short and long reads, as well as variant calling and metaBat
* [hic-map_and_vcf.nf](docs/hic-map_and_vcf.md) - for hi-c reads processing, deduplication, mapping and variant calling (need assembled contigs as input)
* [contigs_annotate.nf](docs/contigs_annotate.md) - for annotating the contigs by mash-screen chromosomes and plasmids database. AMR gene/pMLST gene screening are also carried out with ABRicate
* `dl_fasta.nf` - for downloading genome fasta files from ncbi using list of ncbi accession number as input (e.g. `plsdb_acc.tsv` and `complete_assembly_enterobacteriaceae_chromosome_acc.tsv`). This script is used to download chromosomes and plasmids record for the mash screen by `contigs_annotate.nf`.

#### supplementary scripts (need to incorporate into nextflow steps in the future):
* `chr_mash_result_edit.R` - Used for processing output file from `contigs_annotate.nf`. For joining chromosome mash result with chromosome database info pulled from ncbi (`data/input_files/complete_assembly_enterobacteriaceae_chromosome_info.csv`). This script is used to make `data/results/signf_BacChrMashDist_SH_Oct2020.tsv`.
* `plsdb_mash_result_edit.R` - Used for processing output file from `contigs_annotate.nf`. For joining plsdb mash result with plasmid info pulled from PlsDB (`data/input_files/plsdb_20200629.tsv`). This script is used to make `data/results/Signf_PlsDBMashDist_SH_Oct2020.tsv`.
* `plasmid_map.py` - Used to map plasmid-bacteria linkage. Used hi-c bam-file, kraken output file, plasmid output from `plsdb_mash_result_edit.R` as input.
* `plot_plasmid_contact.R` - plot plasmid linkage heatmap using output of `plasmid_map.py`.
* `plot_plasmid_contact_addAMR.R` - plot plasmid linkage heatmap and add AMR annotation. Need R > 4.0 and the ComplexHeatmap package.
* `run_checkm.sh` - run CheckM for quality assessment on the metabat2 output and bin3c output. Dependent on CheckM version < 1.1
* `cleanup.sh` - to clean up the cache and work files generated by nextflow

#### other scripts:
* [assembly-metaflye_and_polish.nf](docs/assembly-metaflye_and_polish.md)
* [shortrd-map_and_bin.nf](docs/shortrd-map_and_bin.md) 
* [chromosome_check.nf](docs/chromosome_check.md)

### `data` contents
-------------------------------------------------------
#### input_files
* runtables that detail the locations and sample ID of the HiC, short-read and long-read dataset for this project on the UTS HPC computer
    * `hicreads.csv`
    * `longreads_*.csv`
    * `shortreads_*.csv`
* `.msh` reference file for running chromosome/plasmid mash screen and their corresponding information from ncbi/PlsDB
    * `complete_assembly_enterobacteriaceae_chromosome_info.csv`
    * `complete_assembly_enterobacteriaceae_chromosome_list.txt` (accession number of the chromosome database)
    * `complete_assembly_enterobacteriaceae_chromosome.msh`
    * `plsdb_20200629.tsv`
    * `plsdb.msh` (too big to upload to github, but this file can be downloaded from PlsDB directly)

#### results
* ABRicate antimicrobial gene screening results (using ncbi and resfinder databases)
* checkM genome quality assessment graphs
* mlplasmid prediction results
* PlsDB plasmid mash screen results
* chromosome mash screen results
* plasmid hic linkage output tables
* plasmid-AMR-bacteria linkage summary table
* plasmid-AMR-bacteria linkage heatmap


### Log
-------------------------------------------------------
* *28 Aug 2020*
    * added workflow for long-read assembly (`script/assembly-metaflye_and_polish.nf`)
    * added workflow for short-read mapping and binning (`script/shortrd-map_and_bin.nf`)

* *1 Sep 2020*
    * modified `shortrd-map_and_bin.nf` to take each of SH/WG set at a time
    * unified input runtables for different scripts
    * added workflow for hiC mapping (`script/hic-map_and_bin3c.nf`)

* *4 Sep 2020*
    * renamed `script` directory to `bin` and made scripts executable
    * added ABRicate antimicrobial resistance genes screening results to `data/results`
    * added qa graphs from `checkM` genome quality assessment to `data/results`

* *9 Sep 2020*
    * added mlplasmid plasmid prediction results to `data/results`
    * added PlsDB sequence Mash plasmid identification results to `data/results`
    * renamed hiC workflow from `hic-map_and_bin3c.nf` to `hic-map_and_vcf.nf`
    * added index and bgzf steps in workflow

* *16 Sep 2020*
    * added new workflow that combined long-read assembly and short-read mapping and binning steps (`bin/assembly_map_vcall.nf`)

* *10 Oct 2020*
    * added new workflow for mash screen contigs against list of bacterial chromosome genomes (`bin/chromosome_check.nf`)
    * list of accession number downloaded for input of `chromosome_check.nf` is in `data/input_files/complete_assembly_enterobacteriaceae_chromosome_list.txt`

* *20 Oct 2020*
    * modified `hic-map_and_vcf.nf` to perform deduplication and remapping when generating bamfiles
    * added R scripts `plsdb_mash_result_edit.R` and `chr_mash_result_edit.R` for reformating mash screen results
    * added R script `plot_plasmid_contact_addAMR.R` for visualizing the plasmid-contact heatmap with AMR and plasmid typing annotation

* *23 Oct 2020*
    * added new workflow for annotating plasmids, chromosomes, AMR genes and pMLST genes on contigs (`bin/contigs_annotate.nf`)
    * updated content table to describe current results
    * added/updated hic plasmid-mapping results to `data/results`
